src/libexpr/eval.cc

```diff
@@ -198,8 +193,11 @@ string showType(const Value & v)
         case tPrimOpApp:
             return fmt("the partially applied built-in function '%s'", string(getPrimOp(v)->primOp->name));
         case tExternal: return v.external->showType();
+        case tThunk: return "a thunk";
+        case tApp: return "a function application";
+        case tBlackhole: return "a black hole";
     default:
-        return showType(v.type);
+        return showType(v.normalType());
```

Comment:
Regarding `showType`, the function `showType(ValueType)` was removed and replaced with `showType(NormalType)`, because it's not really possible to get a `ValueType` anymore. The new `showType(NormalType)` also aborts for thunks, and it's only used on forced values. The `showValue(Value)` function however does need to be able to print thunks, since it's used on non-forced values. So this now handles them before passing the normal type to `showType(NormalType)`